src/AppInstallerRepositoryCore/Microsoft/ARPHelper.cpp

```diff
@@ -8,6 +8,102 @@ namespace AppInstaller::Repository::Microsoft
+            // without dashes and with opposite endian-ness.
+            //
+            // For example
+            //   ARP:          {FECAFEB5-8D0E-4AE4-8FA0-745BAA835C35}
+            //                FECAFEB5 8D0E 4AE4 8F A0 74 5B AA 83 5C 35
+            //                 Part 1   P2   P3  <------ Part 4 ------->
+            //                5BEFACEF E0D8 4EA4 F8 0A 47 B5 AA 38 C5 53
+            //   UpgradeCode:     5BEFACEFE0D84EA4F80A47B5AA38C553
+            //
+            // The conversion can be done by mapping each location in the packed string
+            // to the appropriate location in the unpacked string.
+            constexpr size_t PackedLength = 32;
+            if (packed.length() != PackedLength || !std::all_of(packed.begin(), packed.end(), isxdigit))
+            {
+                return {};
+            }
+
+            // PositionMapping[i] is the position to which the i-th char is mapped
+            // I.e., unpacked[ PositionMapping[i] ] = packed[i]
+            constexpr size_t PositionMapping[PackedLength] =
+            {
+                8,7,6,5,4,3,2,1,
+                13,12,11,10,
+                18,17,16,15,
+                21,20, 23,22,
+                26,25, 28,27, 30,29, 32,31, 34,33, 36,35,
+            };
+
+            std::string unpacked("{00000000-0000-0000-0000-000000000000}");
+            for (size_t i = 0; i < PackedLength; ++i)
+            {
+                unpacked[PositionMapping[i]] = packed[i];
+            }
+
+            return unpacked;
+        }
+
+        // Gets a mapping from ProductCode to UpgradeCode for MSI packages.
+        std::map<std::string, std::string> GetUpgradeCodes()
+        {
+            // The UpgradeCode is not stored in the ARP registry keys, so we have to get it separately.
+            // We could use MsiGetProductProperty or MsiGetProperty from the MSI API to query it,
+            // but it is very slow.
+            //
+            // The UpgradeCode is also stored in the registry under
+            //   HKLM\SOFTWARE\Microsoft\Windows\CurrentVersion\Installer\UpgradeCodes
+            // (Note that this key is not documented, so it is possible that it will change but very unlikely...)
+            //
+            // Under 'UpgradeCodes' there is one key for each upgrade code, and each upgrade code key
+            // contains the product code as a value. All the upgrade codes and product codes are GUIDs,
+            // but represented in an unusual way - see TryUnpackUpgradeCodeGuid()
+
+            AICLI_LOG(Repo, Info, << "Reading MSI UpgradeCodes");
+            std::map<std::string, std::string> upgradeCodes;
+
+            // There is no UpgradeCodes key on the x86 view of the registry
+            Registry::Key upgradeCodesKey = Registry::Key::OpenIfExists(HKEY_LOCAL_MACHINE, "SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\Installer\\UpgradeCodes", 0, KEY_READ | KEY_WOW64_64KEY);

```

Comment:
Small important change since the review to fix test failure.

Reply:
It does make sense that these would only be stored in the OS architecture.